+sw_tests/+unit_tests/unittest_spinw.m

```diff
@@ -0,0 +1,119 @@
+                actual_value = actual_obj.(field{:});
+                if isstruct(expected_value)
+                    testCase.verify_obj(expected_value, actual_value, rel_tol, abs_tol);
+                else
+                    testCase.verify_val(expected_value, actual_value, rel_tol, abs_tol, field{:});
+                end
+            end
+        end
+        function verify_val(testCase, expected_val, actual_val, rel_tol, abs_tol, field)
+            if nargin < 5
+                abs_tol = 0;
+            end
+            if nargin < 4
+                rel_tol = 1e-10;
+            end
+            import matlab.unittest.constraints.IsEqualTo
+            import matlab.unittest.constraints.RelativeTolerance
+            import matlab.unittest.constraints.AbsoluteTolerance
+            theseBounds = RelativeTolerance(rel_tol) | AbsoluteTolerance(abs_tol);
+            testCase.verifyThat(actual_val, IsEqualTo(expected_val, 'Within', theseBounds), field);
+        end
+    end
+    methods (Test)
+        function test_spinw_no_input(testCase)
+            % Tests that if spinw is called with no input, a default spinw
+            % object is created
+            expected_spinw = testCase.load_spinw('spinw_default.mat');
+            actual_spinw = spinw;
+            testCase.verify_obj(expected_spinw, actual_spinw);
+        end
+        function test_spinw_from_struct_input(testCase, spinw_struct_input)
+            % Tests that if spinw is called with struct input, the relevant
+            % fields are set
+            expected_spinw = testCase.load_spinw(spinw_struct_input{2});
+            actual_spinw = spinw(spinw_struct_input{1});
+            testCase.verify_obj(expected_spinw, actual_spinw);
+        end
+        function test_spinw_from_spinw_obj(testCase)
+            expected_spinw = testCase.load_spinw('spinw_from_struct_lat224.mat');
+            actual_spinw = spinw(expected_spinw);
+            % Ensure creating from a spinw obj creates a copy
+            assert(actual_spinw ~= expected_spinw);
+            testCase.verify_obj(expected_spinw, actual_spinw);
+        end
+        function test_spinw_from_figure(testCase, spinw_figure_input)
+            expected_spinw = testCase.load_spinw(spinw_figure_input{2});
+            figure = testCase.load_figure(spinw_figure_input{1});
+            actual_spinw = spinw(figure);
+            % Ensure creating from a figure creates a copy
+            assert(actual_spinw ~= expected_spinw);
+            testCase.verify_obj(expected_spinw, actual_spinw);
+            close(figure);
+        end
+        function test_spinw_from_incorrect_figure(testCase)
+            fig = figure('visible', 'off');
+            testCase.verifyError(@() spinw(fig), 'spinw:spinw:WrongInput');
+            close(fig);
+        end
+        function test_spinw_from_file(testCase, spinw_file_input)
+            fname = fullfile(testCase.get_unit_test_dir(), 'cifs', spinw_file_input{1});
```

Comment:
I agree having the input structs at the top means you have to keep scrolling up/down to read the input. Is there any reason why we need to have the input as a property of the class, and not hard-code it in the call to the function? For example
`fname = fullfile(testCase.get_unit_test_dir(), 'cifs', 'YFeO3_mcphase.cif');`
if two tests were always intended to use the same file we could then make the filename a property instead of the whole input? Maybe that's poor form?

Reply:
A test is run for every entry in each `TestParameter` e.g. `spinw_figure_input` has length 2 so `test_spinw_from_figure` is called twice, once with `{'default_structure.fig', 'spinw_default.mat'}` and once with `{'afm_chain_spec.fig', 'spinw_afm_chain.mat'}`. If we hard code the values we'd have to write 2 tests and end up duplicating a lot more code. https://uk.mathworks.com/help/matlab/matlab_prog/create-basic-parameterized-test.html

Repeated filenames aren't made a property as I'd like to be able to add more test cases easily (by just adding another entry in `spinw_figure_input`) if we find we need to cover more edge cases for example.

Sorry if I've misunderstood what you meant here! But I do agree the parameters being all at the top is not so good... in Pytest they're just above the function and it's a lot easier to read, so maybe there's a better Matlab-style way to doing this as currently I'm just using my Pytest brain.

Reply:
Ah I see thanks, sorry I'm unfamiliar with MATLAB testing framework. In mantid it is normally preferred to have smaller similar tests so that it is obvious which use case breaks the test - e.g.
`test_spinw_figure_import_with_default_fig`
`test_spinw_figure_import_with_afm_chain`
with a helper function to handle the shared code between the two - but maybe MATLAB makes this difficult to do this in the same file! Whatever works, I'm happy as is, now I understand how the arguments are being passed!

Reply:
Ah that is another valid way of doing things! You've got me looking up the advantages and disadvantages of parametrised tests now, that would be one option to avoid the inputs being so far away from the actual test...